### PR TITLE
[JSON-API] Backport of connection pool for 1.11.x , configurable db options needs forward porting

### PR DESCRIPTION
**Pull request by akshayshirahatti-da**
Backport of Hikari connection pool for 1.11.x plus addition of a configurable pool size parameter to jdbcConfig.

Forward porting of configurable db options for `main` is on [#11621](https://github.com/digital-asset/daml/pull/11621)

CHANGELOG_BEGIN
[JSON-API] Backport of Hikari connection pool for 1.11.x plus addition of a configurable pool size parameter to jdbcConfig.
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/Config.scala [hunk 1]
```diff
@@ -102,15 +108,24 @@ private[http] final case class JdbcConfig(
     user: String,
     password: String,
     createSchema: Boolean = false,
+    minIdle: Int = JdbcConfig.MinIdle,
```
Comment (cocreature) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/Config.scala:111]: so this part is different from the strict backport right? In the future, I’d recommend separating the backport from additional changes to ease review

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/Config.scala [hunk 2]
```diff
@@ -134,6 +151,7 @@ private[http] object JdbcConfig extends ConfigCompanion[JdbcConfig]("JdbcConfig"
     "<user>",
     "<password>",
     "<true|false>",
+    "<poolSize>",
```
Comment (cocreature) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/Config.scala:154]: would it make sense to make the other options also configurable?

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/Main.scala [hunk 3]
```diff
@@ -82,7 +82,7 @@ object Main extends StrictLogging {
       case _ =>
     }
 
-    val serviceF: Future[HttpService.Error \/ ServerBinding] =
+    val serviceF: Future[HttpService.Error \/ (ServerBinding, Option[ContractDao])] =
```
Comment (cocreature) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/Main.scala:85]: maybe worth having an `httpservice` class that contains the server binding & optional dao and abstracts away the details of closing? Not as part of the backport but separately

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala [hunk 4]
```diff
@@ -28,6 +43,26 @@ class ContractDao private (xa: Connection.T)(implicit val jdbcDriver: SupportedJ
 
   def isValid(timeoutSeconds: Int): IO[Boolean] =
     fconn.isValid(timeoutSeconds).transact(xa)
+
+  @SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
```
Comment (cocreature) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala:47]: maybe `discard` instead?

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/Config.scala [hunk 5]
```diff
@@ -102,15 +108,24 @@ private[http] final case class JdbcConfig(
     user: String,
     password: String,
     createSchema: Boolean = false,
+    minIdle: Int = JdbcConfig.MinIdle,
+    connectionTimeout: Long = JdbcConfig.ConnectionTimeout,
+    idleTimeout: Long = JdbcConfig.IdleTimeout,
+    poolSize: Int = ConnectionPool.PoolSize.Integration,
```
Comment (S11001001) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/Config.scala:114]: This default is dangerous.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala [hunk 6]
```diff
@@ -266,3 +304,46 @@ object ContractDao {
     val SqlState = "STALE_OFFSET_EXCEPTION"
   }
 }
+
+/*
+  TODO below values are hardcoded for now, refactor to be picked up as cli flags/ props later.
```
Comment (cocreature) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala:309]: does this comment still apply given that now it’s configurable?

**Review by cocreature — APPROVED** [review]
Nice! Thanks for tackling this so quickly

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/Config.scala [hunk 7]
```diff
@@ -120,12 +135,14 @@ private[http] object JdbcConfig extends ConfigCompanion[JdbcConfig]("JdbcConfig"
...
       s"${indent}password -- database user password,\n" +
       s"${indent}createSchema -- boolean flag, if set to true, the process will re-create database schema and terminate immediately.\n" +
+      s"${indent}poolSize -- int value, specifies the max pool size for the database connection pool.\n" +
       s"${indent}Example: " + helpString(
         "org.postgresql.Driver",
         "jdbc:postgresql://localhost:5432/test?&ssl=true",
         "postgres",
         "password",
         "false",
+        "8",
```
Comment (S11001001) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/Config.scala:145]: Unneeded magic, a good constant is available.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala (hunk 6 above)
Comment (S11001001) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala:309]: Done now, right?

Comment (akshayshirahatti-da) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala:309]: nope will cleanup

**Review by S11001001 — COMMENTED** [review]
Please retitle; if we think this PR is only a backport, that will constitute a regression in `main`. (In other words, `needs-forward-port`)

**Pull request merged**